src/string/extglobmatch.c

```diff
@@ -0,0 +1,276 @@
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include "extglobmatch.h"
+#include "brz.h"
+
+#define cat_one()    do{                                    \
+                        if (atoms > 1) {                    \
+                            --atoms;                        \
+                            glob *b = *--frag;              \
+                            glob *a = *--frag;              \
+                            *frag++ = glob_seq(a,b);             \
+                            free_glob(a);                   \
+                            free_glob(b);                   \
+                        }                                   \
+                    } while(0)
+
+#define cat_all()    do{                                    \
+                          while (--atoms > 0) {             \
+                              glob *b = *--frag;            \
+                              glob *a = *--frag;            \
+                              *frag++ = glob_seq(a,b);           \
+                              free_glob(a);                 \
+                              free_glob(b);                 \
+                          }                                 \
+                    }while(0)
+
+static glob* match_glob = NULL;
+
+static void cleanup(glob **start, glob **end) {
+    while (start != end)
+    {
+        free_glob(*start);
+        ++start;
+    }
+}
+
+static void print_error(char const c) {
+    fprintf(stderr, "pattern error: Unexpected character '%c'\n", c);
+}
+
+static glob* handle_operator(char op, glob const* g)
+{
+    glob* tmp = NULL;
+    switch (op) {
+        case '!': tmp = glob_not(g);
+                  break;
+        case '*': tmp = glob_star(g);
+                  break;
+        case '+': tmp = glob_plus(g);
+                  break;
+        case '@': tmp = glob_at(g);
+                  break;
+        case '?': tmp = glob_opt(g);
+                  break;
+    }
+    return tmp;
+}
+
+#define push_frag(f) *frag++ = f
+#define pop_frag() *--frag
+#define ARRAY_SIZE 100
+
+static glob *compile(char const *pat) {
+    glob *frags[ARRAY_SIZE]={0};
+    glob **frag=frags;
+    struct paren {
+        int alter;
+        int atoms;
+        char op;
+    } paren[ARRAY_SIZE], *pa;
+
+    int alter = 0;
+    int atoms = 0;
+    pa = paren;
+
+    if (!pat) {
```

Comment:
Don't bother checking for a `NULL` pattern here, it should be a precondition. Use `assert(pat);`.